var/spack/repos/builtin/packages/autoconf/package.py

```diff
@@ -35,7 +35,9 @@ class Autoconf(Package):
     version('2.69', '82d05e03b93e45f5a39b828dc9c6c29b')
     version('2.62', '6c1f3b3734999035d77da5024aab4fbd')
 
-    depends_on('m4', type='build')
+    # Note: m4 is not only a build-time dependency of autoconf. The m4
+    # executable is needed when autoconf runs.
+    depends_on('m4 @1.4.6:')
```

Comment:
`type=("build", "run")` fits your description, default is `type=("build", "link")`.